pulser-simulation/pulser_simulation/simulation.py

```diff
@@ -225,6 +232,97 @@ def set_config(self, cfg: SimConfig) -> None:
+                    f" in the depolarizing probability. p = {4*prob}"
+                    " is too large for realistic results.",
+                    stacklevel=2,
+                )
+
+            k = np.sqrt((prob) * (1 - prob) ** (n - 1))
+            self._collapse_ops = [
+                np.sqrt((1 - 3 * prob) ** n)
+                * qutip.tensor([self.op_matrix["I"] for _ in range(n)])
+            ]
+            self._collapse_ops += [
+                k
+                * (
+                    self.build_operator([("sigma_rr", [qid])])
+                    - self.build_operator([("sigma_gg", [qid])])
+                )
+                for qid in self._qid_index
+            ]
+            self._collapse_ops += [
+                k
+                * (
+                    self.build_operator([("sigma_gr", [qid])]).dag()
+                    + self.build_operator([("sigma_gr", [qid])])
+                )
+                for qid in self._qid_index
+            ]
+            self._collapse_ops += [
+                1j
+                * k
+                * (
+                    self.build_operator([("sigma_gr", [qid])]).dag()
+                    - self.build_operator([("sigma_gr", [qid])])
+                )
+                for qid in self._qid_index
+            ]
+
+        if "gen_noise" in self.config.noise:
+            if self.basis_name == "digital" or self.basis_name == "all":
+                # Go back to previous config
+                self.set_config(prev_config)
+                raise NotImplementedError(
+                    "Cannot include general "
+                    + "noise in digital- or all-basis."
+                )
+
+            # Probability distribution of error occurences
+            n = self._size
+            m = len(self.config.gen_noise_opers)
+            if n > 1:
+                for i in range(1, m):
+                    prob_i = self.config.gen_noise_probs[i]
+                    if prob_i > 0.1:
+                        warnings.warn(
+                            "The general noise model is a first-order"
+                            " approximation in the noise probability."
+                            f"p={prob_i} is large for realistic results.",
+                            stacklevel=2,
+                        )
+                        break
+            # Building collapse operators
+            prob_id = self.config.gen_noise_probs[0]
+            self._collapse_ops += [
+                np.sqrt(prob_id**n)
+                * qutip.tensor([self.op_matrix["I"] for _ in range(n)])
+            ]
+            for i in range(1, m):
+                k = np.sqrt(
+                    self.config.gen_noise_probs[i] * prob_id ** (n - 1)
+                )
+                c_op = k * self.config.gen_noise_opers[i]
+                self._collapse_ops += [
+                    self.build_operator([(c_op, [qid])])
+                    for qid in self._qid_index
+                ]
```

Comment:
I think you could simplify the code by making the `dephasing` and `depolarizing` cases define the proper Kraus operators and then they would always be built here. 
The current implementation introduces a lot of unnecessary duplication.

Reply:
I must remove them from the set of noise models?

Reply:
No no, they are convenient to have. But you could store their noise operators under `gen_noise_opers` so that the collapse operators are built always in the same place

Reply:
Because the only place where collapse_operators are defined is in `dephasing`, `depolarizing` or `gen_noise` mode

Reply:
I think `effective_nosie` is more convenient than `gen_noise`. What do you think?

Reply:
I agree, I prefer `eff_noise` over `gen_noise` too

Reply:
No no, they are convenient to have. But you could store their noise operators under gen_noise_opers so that the collapse operators are built always in the same place// If I do so it will create confusion since I iterate it in a loop to build Kraus operators.

Reply:
I think you are missing the point. In all three cases, all you are doing is taking the probabilities, building the single-qubit Kraus operators and then building the collapse operators from them. The difference in each case is contained in the single-qubit Kraus operators, so the collapse operators could be built from them in one place rather than repeated every time

Reply:
And why not?

Reply:
In either `dephasing` , `depolarizing` or `eff_noise`, i will build Kraus operators then build all collapse operators at once.

Reply:
Exactly, don't you think it's cleaner this way?

Reply:
Since Kraus operators are built only in one of these modes, I need that storage variable to be defined beforehand, why not defining it as an attribute of the class?

Reply:
`self._Kraus_ops`

Reply:
Can't you do it before the first `if`?

Reply:
I think an attribute is not necessary if you build the collapse operators in the same call

Reply:
Ok, I will do it before the first `if`.